console-subscriber/tests/support/subscriber.rs

```diff
@@ -0,0 +1,323 @@
+    // because we will call `connect_with_connector` with a service that
+    // just returns the `DuplexStream`, instead of making an actual
+    // network connection.
+    let endpoint = Endpoint::try_from("http://[::]:6669").expect("Could not create endpoint");
+    let channel = endpoint
+        .connect_with_connector(service_fn(move |_: Uri| {
+            let client = client_stream.take();
+
+            async move {
+                client.ok_or_else(|| {
+                    std::io::Error::new(std::io::ErrorKind::Other, "Client already taken")
+                })
+            }
+        }))
+        .await
+        .expect("client-console error: couldn't create client");
+    test_state.advance_to_step(TestStep::ClientConnected);
+
+    record_actual_tasks(channel, test_state).await
+}
+
+/// Records the actual tasks which are received by the client channel.
+///
+/// Updates will be received until the test state reaches the `TestFinished` step
+/// (indicating that the test itself has finished running), at which point we wait
+/// for a final update before returning all the actual tasks which were recorded.
+///
+/// # Test State
+///
+/// 1. Waits for: `TestFinished`
+async fn record_actual_tasks(
+    client_channel: Channel,
+    mut test_state: TestState,
+) -> Vec<ActualTask> {
+    let mut client = InstrumentClient::new(client_channel);
+
+    let mut stream = match client
+        .watch_updates(tonic::Request::new(InstrumentRequest {}))
+        .await
+    {
+        Ok(stream) => stream.into_inner(),
+        Err(err) => panic!("Client cannot connect to watch updates: {err}"),
+    };
+
+    let mut tasks = HashMap::new();
+
+    let signal_task = ExpectedTask::default().match_name(END_SIGNAL_TASK_NAME.into());
+    let mut signal_task_read = false;
+    while let Some(update) = stream.next().await {
+        let update = update.expect("update stream error");
+
+        if let Some(task_update) = &update.task_update {
+            for new_task in &task_update.new_tasks {
+                let mut actual_task = match new_task.id {
+                    Some(id) => ActualTask::new(id.id),
+                    None => continue,
+                };
+                let name = new_task
+                    .fields
+                    .iter()
+                    .find_map(|field| match field.name.as_ref()? {
+                        console_api::field::Name::StrName(name) if name == "task.name" => {
+                            Some(field.value.clone())
+                        }
+                        _ => None,
+                    })
+                    .flatten();
+                actual_task.name = match name {
+                    Some(Value::DebugVal(value)) => Some(value.clone()),
+                    Some(Value::StrVal(value)) => Some(value.clone()),
+                    _ => None, // Anything that isn't string-like shouldn't be used as a name.
+                };
```

Comment:
i`m normally a big fan of cool iterator tricks, but isn't it simpler to just write something like
```suggestion
                for field in new_task.fields {
                    match field.name.as_ref() {
                        console_api::field::Name::StrName(name) if name == "task.name" => {
                            actual_task.name = match field.value.as_ref() {
                                Value::DebugVal(value) => actual_task.name = Some(value),
                                Value::StrVal(value) => actual_task.name = Some(value),
                                _ => continue;
                            }
                        },
                        _ => {}
                    }
                }
```